From 6f35394823b86d7ab53f9e70821724a1d8b347cd Mon Sep 17 00:00:00 2001
From: Peter Chiu <chui-hao.chiu@mediatek.com>
Date: Thu, 14 Aug 2025 16:12:11 +0800
Subject: [PATCH 117/123] mtk: mt76: mt7996: support sw fast path when wed is
 disabled

When wed is disabled, net_fill_forward_path would still be called
when using sw fast path. If this is not defined or return error,
it cannot use sw fast path. So do not return error and fill in the
required fields.

Signed-off-by: Peter Chiu <chui-hao.chiu@mediatek.com>
---
 mt7996/main.c | 14 +++++++-------
 1 file changed, 7 insertions(+), 7 deletions(-)

diff --git a/mt7996/main.c b/mt7996/main.c
index 6eeb4798..52fb6332 100644
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -2671,7 +2671,6 @@ out:
 	return ret;
 }
 
-#ifdef CONFIG_NET_MEDIATEK_SOC_WED
 static int
 mt7996_net_fill_forward_path(struct ieee80211_hw *hw,
 			     struct ieee80211_vif *vif,
@@ -2679,6 +2678,7 @@ mt7996_net_fill_forward_path(struct ieee80211_hw *hw,
 			     struct net_device_path_ctx *ctx,
 			     struct net_device_path *path)
 {
+#ifdef CONFIG_NET_MEDIATEK_SOC_WED
 	struct mt7996_vif *mvif = (struct mt7996_vif *)vif->drv_priv;
 	struct mt7996_sta *msta = (struct mt7996_sta *)sta->drv_priv;
 	struct mt7996_dev *dev = mt7996_hw_dev(hw);
@@ -2714,10 +2714,8 @@ mt7996_net_fill_forward_path(struct ieee80211_hw *hw,
 	}
 
 	if (!mtk_wed_device_active(wed))
-		return -ENODEV;
+		goto out;
 
-	path->type = DEV_PATH_MTK_WDMA;
-	path->dev = ctx->dev;
 	path->mtk_wdma.wdma_idx = wed->wdma_idx;
 	path->mtk_wdma.bss = mlink->idx;
 	path->mtk_wdma.queue = 0;
@@ -2729,13 +2727,15 @@ mt7996_net_fill_forward_path(struct ieee80211_hw *hw,
 		path->mtk_wdma.amsdu = msta_link->wcid.amsdu;
 	else
 		path->mtk_wdma.amsdu = 0;
-
+out:
+#endif
+	path->type = DEV_PATH_MTK_WDMA;
+	path->dev = ctx->dev;
 	ctx->dev = NULL;
 
 	return 0;
 }
 
-#endif
 
 static int
 mt7996_change_vif_links(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
@@ -3093,8 +3093,8 @@ const struct ieee80211_ops mt7996_ops = {
 	.vif_add_debugfs = mt7996_vif_add_debugfs,
 #endif
 	.set_radar_background = mt7996_set_radar_background,
-#ifdef CONFIG_NET_MEDIATEK_SOC_WED
 	.net_fill_forward_path = mt7996_net_fill_forward_path,
+#ifdef CONFIG_NET_MEDIATEK_SOC_WED
 	.net_setup_tc = mt76_wed_net_setup_tc,
 #endif
 	.event_callback = mt7996_event_callback,
-- 
2.45.2

